### PR TITLE
Avoid using barrier_sync as much as possible

### PR DESCRIPTION
**Pull request by naoyam**
This is a follow-up PR of #351. Extend the same approach to all runtime functions that use `block_sync`.

`barrier_sync` is still used in about 100 C++ tests out of nearly 1000 tests. H100 is now green.

### CODE REVIEW

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
Here's the overall performance change with the use of `syncthreads`. The baseline always uses `barrier_block`, whereas the `aligned` case uses `syncthreads` when possible.

image

Overall, not much difference, but there are some instances that benefit from the aligned sync. Here's the histogram:

image

Note that there are two instances that have around 7-8 % of regressions. One of them seems just a random noise as I can't reproduce. Another one, `NvFuserScheduler_LayerNormFusedOp_fp16/8192/16384`, is reproducible and has around 10% regression, but its bandwidth is just 280 GB/s even with the current main:

```
NvFuserScheduler_LayerNormFusedOp_fp16___GRAPH/NvFuserScheduler_LayerNormFusedOp_fp16/8192/16384/manual_time       1733 us         1808 us          808 bytes_per_second=288.567G/s Red On Fastest Dim // Persistent Kernel //  // Iteration Domain:  // Inner Reduction Domain: cross block reduction / pad to warp / persistent batch - 8 / vectorize / factor 8/Launch_Parameters[block(1/1/256)/grid(1/1/8192)/1024]
```

There seems to be a performance problem no matter which synchronization method is used. CC: @liqiangxl

File: csrc/codegen.cpp [hunk 1]
```diff
@@ -1057,23 +1074,24 @@ class CudaKernelGenerator : private kir::ConstIrVisitor {
...
     const auto data_type = stmt->out()->dtype();
-    indent() << "broadcast::blockBroadcast<" << flags_str.str() << ">(\n";
-    indent() << kTab << gen(stmt->out()) << ",\n";
-    indent() << kTab << gen(stmt->in()) << ",\n";
-    indent() << kTab << "static_cast<" << data_type << "*>(shared_mem),\n";
+
+    ArgumentBuilder func_args;
+    func_args.arg(gen(stmt->out()));
+    func_args.arg(gen(stmt->in()));
+    func_args.arg(genCall("static_cast", genPtrType(data_type), "shared_mem"));
```
Comment (zasdfgbnm) [inline csrc/codegen.cpp:1088]:
```suggestion
    func_args.arg(genStaticCast(genPtrType(data_type), "shared_mem"));
```

File: csrc/codegen.cpp [hunk 2]
```diff
@@ -1637,7 +1664,7 @@ class CudaKernelGenerator : private kir::ConstIrVisitor {
 
     const auto reduction_name = genFusedReductionName(out->view());
 
-    // template <typename Func, typename... Types>
+    // template <bool ALIGNED, typename Func, typename... Types>
```
Comment (zasdfgbnm) [inline csrc/codegen.cpp:1667]: nit:
```suggestion
    // template <bool Aligned, typename Func, typename... Types>
```

File: csrc/codegen.cpp [hunk 3]
```diff
@@ -2370,7 +2407,7 @@ class CudaKernelGenerator : private kir::ConstIrVisitor {
...
```
Comment (zasdfgbnm) [inline csrc/codegen.cpp:2410]: nit
```suggestion
    // template <bool Aligned, typename Func, typename... Types>
```

File: runtime/block_welford_outer.cu [hunk 4]
```diff
@@ -56,7 +56,7 @@ namespace impl {
 // registers than just returing the output. Results would vary
 // depending on compiler versions, but it seems safer to return outputs
 // as a new value.
-template <int NumVals, typename DataType, int BDIMX, int BDIMY>
+template <bool ALIGNED, int NumVals, typename DataType, int BDIMX, int BDIMY>
```
Comment (zasdfgbnm) [inline runtime/block_welford_outer.cu:59]: nit: use variable names of the same style
```suggestion
template <bool ALIGNED, int NUM_VALS, typename DATA_TYPE, int BDIMX, int BDIMY>
```

Comment (zasdfgbnm) [reply runtime/block_welford_outer.cu:59]: Hmm, maybe not a blocker for this PR, but naming variables with different style is very ugly, and we have a lot of it....

File: csrc/codegen.cpp [hunk 5]
```diff
@@ -1352,11 +1365,88 @@ class CudaKernelGenerator : private kir::ConstIrVisitor {
     }
   }
 
+  void genBlockWelford(const WelfordOp* wop) {
```
Comment (zasdfgbnm) [inline csrc/codegen.cpp:1368]: There was a `if (has_block_reduce)` guarding this, but now that check is gone. Is this the expected behavior?

Comment (naoyam) [reply csrc/codegen.cpp:1368]: Thanks for catching this. Yes, it should be guarded.

File: runtime/block_welford_outer.cu (hunk 4 above)
Comment (naoyam) [reply runtime/block_welford_outer.cu:59]: Yeah, right, haven't paid much attention to it. I'd prefer the camel case like `Aligned`, `NumVals`, etc. That's generally the case in our `csrc` codebase. I'll change `ALIGNED` to `Aligned`.

**Review by zasdfgbnm — APPROVED** [review]

**Pull request merged**